src/ghost_borrow_mut.rs

```diff
@@ -247,6 +246,63 @@ generate_public_instance!(a, b, c, d, e, f, g, h, i, j, k, l ; T0, T1, T2, T3, T
 //  Implementation
 //
 
+//  This trait is a pure implementation detail, which only needs to be public because it must be part of the signature
+//  of public functions/traits.
+mod sealed {
+
+#[derive(Clone, Copy, Debug)]
+pub enum Span {
+    Single(*const u8),
+    Slice(*const u8, *const u8),
+}
+
+#[doc(hidden)]
+pub trait GetSpan {
+    //  #   Safety
+    //
+    //  At the moment, it is assumed that the address of a fat pointer always points to the first byte of the actual data.
+    //  If this does not hold, the span returned will not help enforce the necessary invariants.
+    fn get_span(&self) -> Span;
+}
+
+impl<T: ?Sized> GetSpan for T {
+    default fn get_span(&self) -> Span { Span::Single(self as *const T as *const u8) }
+}
+
+impl<T> GetSpan for [T] {
+    fn get_span(&self) -> Span {
+        if self.len() <= 1 {
+            Span::Single(self.as_ptr() as *const u8)
+        } else {
+            let start = self.as_ptr() as *const u8;
+            let length = self.len() * core::mem::size_of::<T>();
```

Comment:
If T is a ZST, this creates an empty slice, which then doesn't overlap anything, including itself.
Therefore, if we have two of a ZST which isn't supposed to be copied (like `GhostToken` itself) we would be able to get however many `&mut` references to it as we wished, which is unsafe.

Reply:
That's an excellent point. `length` will need to be at least one to avoid that.